### PR TITLE
Test and fix polar.py, fftutils.py; refactor FRC/FSC to use fftutils

### PR DESCRIPTION
**Pull request by sakoho81**
### Bug fixes
- **`polar.py`**: `generate_polar_coordinate_grid` had an odd-shape centering bug — `-shape//2` (floor-division of negative) produced asymmetric axes for odd dimensions like `[-3,-2,-1,0,1]` instead of the centered `[-2,-1,0,1,2]`. Fixed to use `-floor(shape/2.0)`, matching `SimplePolarIndexer`'s convention.
- **`fftutils.py`**: `ifft()` used `fftshift` before the inverse FFT, which is not self-inverse for odd-shaped arrays. Fixed to `ifftshift` — roundtrip now correct for both even and odd shapes (was diff ~1.8, now ~1e-16).
- **`fftutils.py`**: `fft()` had `*kwargs` (positional-varargs named like kwargs) — fixed to `**kwargs`.

### Refactoring (safe, verified equivalents)
- **`fourier_ring_correlation.py`**: 2 raw `fftshift(fft2(...))` calls → `fftutils.fft(image, window=None)`
- **`fourier_shell_correlation.py`**: 3 raw `fftshift(fftn(...))` calls → `fftutils.fft(image, window=None)`
- All 5 call sites verified as exact equivalents (windowing + zero-padding happens upstream)

### Code cleanup
- **`polar.py`**: bare asserts → TypeError/ValueError, type annotations, concise docstrings
- **`fftutils.py`**: bare `assert isinstance(image, Image)` ×3 → TypeError, type annotations, concise docstrings

### TODO breadcrumbs (no behavior change)
- `processing/image.py::translate_image()`: flag no-shift-before-ifftn asymmetry
- `registration/stack.py`, `registration.py`: flag unshifted fourier_shift convention
- `ism/helpers.py`: flag manual centered coordinate grid (dup of polar.py)
- `image_quality/filters.py`: flag hand-rolled physical-frequency scaling (2 sites)

### Tests
- **`tests/data/coordinates/test_polar.py`** (16 tests): grid centering even/odd, spacing scaling, indexer shapes (2D/3D meshgrid), radius center/corner values, shape validation errors, low/high-pass mask correctness
- **`tests/processing/test_fftutils.py`** (19 tests): fft/ifft roundtrip even/odd/3D, constant-image centering (DC at center), gaussian self-duality, interpolation shape scaling, windowing branches, unknown window error, ideal/butterworth/gaussian filter validation (shape, spacing, threshold errors, type errors), frequency-domain indexer shape

### Full test suite: 254 passing

### CODE REVIEW

File: miplib/data/coordinates/polar.py [hunk 1]
```diff
@@ -1,66 +1,49 @@
...
-        )
-        assert 1 < len(shape) < 4
+    The ``r`` attribute contains radial distances from the array center,
+    which can be indexed with ``[radius]`` to produce boolean masks.
+    """
 
-        # Create Fourier grid
-        axes = (np.arange(-np.floor(i / 2.0), np.ceil(i / 2.0)) for i in shape)
+    def __init__(self, shape: tuple[int, ...]) -> None:
+        if not isinstance(shape, (tuple, list)):
```
Comment (sakoho81) [inline miplib/data/coordinates/polar.py]: do we really care about this?

File: miplib/data/coordinates/polar.py [hunk 2]
```diff
@@ -1,66 +1,49 @@
...
+            raise TypeError(
+                f"shape must be a tuple or list, got {type(shape).__name__}"
+            )
+        if len(shape) not in (2, 3):
```
Comment (sakoho81) [inline miplib/data/coordinates/polar.py:23]: I suppose it could also be 1? Does polar indexing make sense in higher dimensions? just wondering how much we want to clamp this

File: miplib/processing/fftutils.py [hunk 3]
```diff
@@ -1,161 +1,103 @@
...
-    elif window == "tukey":
-        array = windowing.apply_tukey_window(array, *kwargs)
+def fft(
+    array: np.ndarray,
+    interpolation: float = 1.0,
+    window: str | None = "tukey",
+    **kwargs: Any,
+) -> np.ndarray:
+    """Forward FFT with optional zero-padding interpolation and windowing."""
+    if window == "tukey":
```
Comment (sakoho81) [inline miplib/processing/fftutils.py]: lets add an enum for the available windows

Comment (sakoho81) [reply miplib/processing/fftutils.py]: could also be a mapping, to avoid this if/elif chain here. Not really a concern of the fft method

File: miplib/processing/fftutils.py [hunk 4]
```diff
@@ -1,161 +1,103 @@
...
-    """
-    An ideal high/low pass frequency domain noise filter.
-    :param image: an Image object
-    :param threshold: threshold value [0,1], where 1 corresponds
-    to the maximum frequency.
-    :param kind: filter type 'low' for low-pass, 'high' for high pass
-    :return: returns the filtered Image.
-    """
-    assert isinstance(image, Image)
+def ideal_fft_filter(image: Image, threshold: float, kind: str = "low") -> Image:
```
Comment (sakoho81) [inline miplib/processing/fftutils.py:48]: why does this need to be an image?

File: miplib/processing/fftutils.py [hunk 5]
```diff
@@ -1,161 +1,103 @@
...
-           size of the filter
-       cutoff : float
-           relative cutoff frequency of the filter (0 - 1.0)
-       n : int, optional
-           order of the filter, the higher n is the sharper
-           the transition is.
-    :Returns:
-       numpy.ndarray:  filter kernel in 2D centered
-    """
+def gaussian_fft_filter(image: Image, threshold: float) -> Image:
```
Comment (sakoho81) [inline miplib/processing/fftutils.py:93]: same for others? Is it just that the Image object loses its spacing if not explicitely cast?

File: tests/processing/test_fftutils.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    """Gaussian self-duality: |FFT| peaks at center, decays radially."""
+    f = fft(gaussian_2d, window=None)
+    mag = np.abs(f)
+    center = (mag.shape[0] // 2, mag.shape[1] // 2)
+    peak_idx = np.unravel_index(np.argmax(mag), mag.shape)
+    assert peak_idx == center
+    assert mag[center] > mag[center[0] + 5, center[1]]
+
+
+def test_fft_interpolation_doubles_shape():
```
Comment (sakoho81) [inline tests/processing/test_fftutils.py]: we could also check with a gaussian, that it actually doubles the size, from e.g. fwhm

File: tests/processing/test_fftutils.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+
+def test_fft_window_none_preserves_shape():
+    x = np.random.rand(16, 16)
+    f = fft(x, window=None)
+    assert f.shape == (16, 16)
+    assert f.dtype == np.complex128
+
+
+def test_fft_window_hamming_preserves_shape():
```
Comment (sakoho81) [inline tests/processing/test_fftutils.py]: we could check what this produces, e.g. when applied on ones?

File: tests/processing/test_fftutils.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        ideal_fft_filter(np.ones((8, 8)), threshold=0.5)
+
+
+def test_butterworth_fft_filter_shape_and_spacing(camera_image):
+    filtered = butterworth_fft_filter(camera_image, threshold=0.3)
+    assert filtered.shape == camera_image.shape
+    assert filtered.spacing == camera_image.spacing
+
+
+def test_butterworth_fft_filter_invalid_threshold(camera_image):
```
Comment (sakoho81) [inline tests/processing/test_fftutils.py:280]: could we use some simple test patterns here, to actually verify the function? e.g. rectangles, waves etc.?

**Review by sakoho81 — COMMENTED** [review]
## Review responses

### generate_polar_coordinate_grid
Decided to keep it as a useful utility for physical-unit coordinate grid generation (vs SimplePolarIndexer which works in pixel indices). Added TODO breadcrumbs in `ism/helpers.py` and `image_quality/filters.py` where similar patterns are hand-rolled, so future sessions can refactor onto this.

### Shape constraint (2D/3D only)
Kept the original constraint. 1D doesn't produce meaningful polar indexing (radius would be a single point). 4D+ is not a microscopy use case. The original code explicitly guarded for `1 < len(shape) < 4`.

### Window enum → Literal type
Added `Literal["tukey", "hamming"]` for the `window` parameter in `fft()`. This gives mypy type checking without adding enum boilerplate. The `"bad"` test case uses `# type: ignore[arg-type]`.

### Window mapping → dict
Replaced the if/elif with `_WINDOW_FUNCS: dict[str, Callable[..., np.ndarray]]` mapping. Added `Callable[..., np.ndarray]` type annotation to keep mypy happy.

### Image requirement for filters
All three filter functions capture `image.spacing` and use it to wrap the output in `Image(result, spacing)`. Without Image input the output would be a plain ndarray with no spacing metadata — the spacing flows through the pipeline by design. The isinstance check + TypeError replaces the old bare assert.

### Same for butterworth/gaussian
Yes — all three follow the same pattern: extract spacing → FFT → apply frequency mask → inverse FFT → re-wrap in `Image(..., spacing)`.

### Interpolation FWHM test
Added `test_fft_interpolation_increases_frequency_resolution` verifying `interpolation=2.0` doubles the FFT output shape (32×32 → 64×64) which enables finer frequency-domain sampling.

### Filter-on-constant tests
Added `test_ideal_filter_on_constant_low_pass` and `test_ideal_filter_on_constant_high_pass`, plus `test_butterworth_filter_on_constant_returns_constant` and `test_gaussian_filter_on_constant_returns_constant`. Constant images have only DC energy — low-pass preserves, high-pass zeros, butterworth/gaussian preserve.

### Test patterns (sine grating)
Added `test_ideal_fft_filter_low_pass_reduces_variance` (low-pass on sine grating → variance drops significantly because the sinusoid is at frequency 0.3, above cutoff 0.1*nnq) and `test_ideal_fft_filter_high_pass_on_sine` (high-pass preserves the oscillation energy). Using `tests.conftest.sine_grating` generator.

**Review by sakoho81 — COMMENTED** [review]
Re: shape type constraint — agreed, any container with a length works. Removed the isinstance(tuple, list) check entirely. Now only validates `len(shape) in (2, 3)`. Removed the corresponding TypeError test. Pushed in the latest commit.

**Review by sakoho81 — COMMENTED** [review]
Re: richer test patterns — added `step_edge`, `bin_aligned_sine`, `two_frequency_signal` to `conftest.py`. Replaced the shallow shape-only tests with:

- **Windowing**: window equivalence (hamming), alpha propagation (tukey), spectral leakage reduction (non-bin-aligned sine shows windowing suppresses sidelobes)
- **Step-edge ringing**: ideal filter produces overshoot >1.0 (Gibbs), Gaussian has no overshoot, Butterworth ringing increases with order n
- **Non-constant filter tests**: all 4 filters verified to actually change non-constant input (contra all-pass). Constant identity tests kept as complementary edge case.
- **Interpolation**: consolidated shape+FWHM into one test
- Note: direct passband/stopband FFT magnitude tests were blocked by the `np.abs(.real)` post-processing in the filter functions — switched to spatial-domain assertions instead.
